src/hotspot/share/gc/g1/heapRegionRemSet.inline.hpp

```diff
@@ -55,10 +55,64 @@ void HeapRegionRemSet::set_state_complete() {
   _state = Complete;
 }
 
+template <typename Closure>
+class G1ContainerCardsOrRanges {
+  Closure& _iter;
+  uint _region_idx;
+  uint _offset;
+
+public:
+  G1ContainerCardsOrRanges(Closure& iter, uint region_idx, uint offset) : _iter(iter), _region_idx(region_idx), _offset(offset) { }
+
+  bool start_iterate(uint tag) {
+    return _iter.start_iterate(tag, _region_idx);
+  }
+
+  void operator()(uint card_idx) {
+    _iter.do_card(card_idx + _offset);
+  }
+
+  void operator()(uint card_idx, uint length) {
+    _iter.do_card_range(card_idx + _offset, length);
+  }
+};
+
+template <typename Closure, template <typename> class CardOrRanges>
+class G1HeapRegionRemSetMergeCardIterator : public G1CardSet::G1CardSetPtrIterator {
+  G1CardSet* _card_set;
+  Closure& _iter;
+  uint _log_card_regions_per_region;
+  uint _card_regions_per_region_mask;
+  uint _log_card_region_size;
+
+public:
+
+  G1HeapRegionRemSetMergeCardIterator(G1CardSet* card_set,
+                                      Closure& iter,
+                                      uint log_card_regions_per_region,
+                                      uint log_card_region_size) :
+    _card_set(card_set),
+    _iter(iter),
+    _log_card_regions_per_region(log_card_regions_per_region),
+    _card_regions_per_region_mask((1 << log_card_regions_per_region) - 1),
+    _log_card_region_size(log_card_region_size) {
+  }
+
+  void do_cardsetptr(uint region_idx, size_t num_occupied, G1CardSet::CardSetPtr card_set) override {
+    CardOrRanges<Closure> cl(_iter,
+                             region_idx >> _log_card_regions_per_region,
+                             (region_idx & _card_regions_per_region_mask) << _log_card_region_size);
+    _card_set->iterate_cards_or_ranges_in_container(card_set, cl);
```

Comment:
`region_idx` is actually `card_region_idx`, right? If so, `card_region_idx >> _log_card_regions_per_region` ==> `region_idx` (the second arg of `G1ContainerCardsOrRanges` constructor) makes sense to me.

Reply:
Yes, that is the unchanged "region" that the hash table for the containers sees, which is the `card_region_idx` at this level.